build_tools/github_actions/runner/instance_deleter/main.py

```diff
@@ -0,0 +1,211 @@
+import requests
+from google.api_core import exceptions
+from google.auth import transport
+from google.cloud import compute
+from google.oauth2 import id_token
+
+AUTH_HEADER_PREFIX = "Bearer "
+
+instances_client = compute.InstancesClient()
+migs_client = compute.RegionInstanceGroupManagersClient()
+session = requests.Session()
+
+print("Server started")
+
+
+def verify_token(token: str) -> dict:
+  """Verify token signature and return the token payload"""
+  request = transport.requests.Request(session)
+  payload = id_token.verify_oauth2_token(token, request=request)
+  return payload
+
+
+def get_region(zone: str) -> str:
+  """Extract region name from zone name"""
+  # Drop the trailing zone identifier to get the region. Yeah it kinda does seem
+  # like there should be a better way to do this...
+  region, _ = zone.rsplit("-", maxsplit=1)
+  return region
+
+
+def get_name_from_resource(resource: str) -> str:
+  """Extract just the final name component from a fully scoped resource name."""
+  _, name = resource.rsplit("/", maxsplit=1)
+  return name
+
+
+def get_from_items(items: compute.Items, key: str):
+  # Why would the GCP Python API return something as silly as a dictionary?
+  return next((item.value for item in items if item.key == key), None)
+
+
+@functions_framework.http
+def delete_self(request):
+  """HTTP Cloud Function.
+    Args:
+        request (flask.Request): The request object.
+        <https://flask.palletsprojects.com/en/1.1.x/api/#incoming-request-data>
+    Returns:
+        The response text, or any set of values that can be turned into a
+        Response object using `make_response`
+        <https://flask.palletsprojects.com/en/1.1.x/api/#flask.make_response>.
+    Note:
+        For more information on how Flask integrates with Cloud
+        Functions, see the `Writing HTTP functions` page.
+        <https://cloud.google.com/functions/docs/writing/http#http_frameworks>
+  """
+  if request.method != "DELETE":
+    return flask.abort(
+        400, f"Invalid method '{request.method}'. Only DELETE is supported.")
+
+  # No path is needed, since the token contains all the information we need.
+  if request.path != "/":
+    return flask.abort(
+        400,
+        f"Invalid request path '{request.path}'. Only root path is valid).")
+
+  auth_header = request.headers.get("Authorization")
+  if auth_header is None:
+    return flask.abort(401, "Authorization header is missing")
+  if not auth_header.startswith(AUTH_HEADER_PREFIX):
+    return flask.abort(
+        401, f"Authorization header does not start with expected string"
+        f" '{AUTH_HEADER_PREFIX}'.")
+
+  token = auth_header[len(AUTH_HEADER_PREFIX):]
+
+  try:
+    # We don't verify audience here because Cloud IAM will have already done so
+    # and jwt's matching of audiences is exact, which means trailing slashes or
+    # http vs https matters and that's pretty brittle.
+    token_payload = verify_token(token)
+  except ValueError as e:
```

Comment:
Should we catch `exceptions.GoogleAuthError ` here since `verify_oauth2_token` throws that?